examples/adjoints/adjoints_ex5/heatsystem.C

```diff
@@ -62,8 +62,10 @@ void HeatSystem::init_data ()
 
   ZeroFunction<Number> zero;
 
+  // Most DirichletBoundary users should use this style of
+  // constructor, which takes a "locally indexed" functor
   this->get_dof_map().add_dirichlet_boundary
-    (DirichletBoundary (all_bdys, T_only, &zero));
+    (DirichletBoundary (zero, all_bdys, T_only));
```

Comment:
No, it's even more confusing than that.  :-D  The old APIs can also accept references instead of pointers.

Your enum idea sounds like a big improvement over either this confusing disambiguation or my named constructor alternative.